DEVDOCS-5037: [new] GraphQL Cart and Checkout (#1953)
bigcommerce/dev-docs: docs/api-docs/storefront/graphql/graphql-carts-and-checkout.mdx

The GraphQL Storefront API returns the checkout ID and order ID, which you can use to [Get accepted payment methods](/docs/rest-payments/methods#get-accepted-payment-methods) and [Create a payment access token](/docs/rest-payments/tokens#create-payment-access-token). It also returns the customer ID, which you can use to [Get stored payment instruments](/docs/rest-management/customers/customer-stored-instruments#get-stored-instruments). To learn more about authenticating REST endpoints, locate the **Authentication** section at the top of each endpoint, then click **Show Details**.

Gift certificates are a type of payment instrument similar to a debit card, so the Payments API must apply the gift certificate to the checkout.

The API account you use to process payments must have the scopes specified in the [Payments API Overview](/api-docs/store-management/payments-api-overview). 

## Example queries and mutations

### Create a cart using a simple product

<Tabs items={['Request', 'Response']}>
  <Tab>
    ```graphql filename="Example mutation: Create a cart (simple)" showLineNumbers copy 
    # Creates a new cart, adding a simple product.

>     # This will fail in the GraphQL Playground if you already have a cart.

Reviewer: ```suggestion
    # If you already have a cart, this will fail in the GraphQL Playground.
```

    {
      "data": {
        "site": {
          "cart": {
            "entityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d",
            "currencyCode": "USD",
            "isTaxIncluded": false,
            "baseAmount": {
              "currencyCode": "USD",
              "value": 25
            },
            "discountedAmount": {
              "currencyCode": "USD",
              "value": 0
            },
            "amount": {
              "currencyCode": "USD",
              "value": 25
            },
            "discounts": [
              {
                "entityId": "50892aad-8e57-4a86-9668-c1f1ee493cff",
                "discountedAmount": {
                  "currencyCode": "USD",
                  "value": 0
                }
              }
            ],
            "lineItems": {
              "physicalItems": [
                {
                  "entityId": "50892aad-8e57-4a86-9668-c1f1ee493cff",
                  "parentEntityId": null,
                  "variantEntityId": 74,
                  "productEntityId": 111,
                  "sku": "SM13",
                  "name": "[Sample] Smith Journal 13",

>                   "url": "https://cart-graphql-playground.mybigcommerce.com/smith-journal-13/",

Reviewer: ```suggestion
                  "url": "https://cart-graphql-playground.mybigcommerce.com/smith-journal-13/",
```
make this URL fake

>                   "imageUrl": "https://cdn11.bigcommerce.com/s-vjbwqbabp1/products/111/images/371/smithjournal1.1679614893.220.290.jpg?c=1",

Reviewer: ```suggestion
                  "imageUrl": "https://cdn11.bigcommerce.com/s-vjbwqbabp1/products/111/images/371/smithjournal1.1679614893.220.290.jpg?c=1",
```
make this image URL fake

                  },
                  "salePrice": {
                    "currencyCode": "USD",
                    "value": 25
                  },
                  "extendedListPrice": {
                    "currencyCode": "USD",
                    "value": 25
                  },
                  "extendedSalePrice": {
                    "currencyCode": "USD",
                    "value": 25
                  },
                  "isShippingRequired": true,
                  "selectedOptions": [],
                  "giftWrapping": null
                }
              ],
              "digitalItems": [],
              "giftCertificates": [],
              "customItems": [],
              "totalQuantity": 1
            },
            "createdAt": {
              "utc": "2023-06-01T17:45:23Z"
            },
            "updatedAt": {
              "utc": "2023-06-01T17:45:23Z"
            },
            "locale": "en"
          }
        }
      }
    }
    ```
  </Tab>
</Tabs>

### Add cart line items

<Tabs items={['Request', 'Response']}>
  <Tab>
    ```graphql filename="Example mutation: Add cart line items" showLineNumbers copy
    # Adds another line item.

>     # Update the GQL variables with details from your cart first, or it will fail.

Reviewer: ```suggestion
    # Add your cart's ID to the GraphQL variables, or the mutation will fail. 
```

    mutation addCartLineItems($addCartLineItemsInput: AddCartLineItemsInput!) {
      cart {
        addCartLineItems(input: $addCartLineItemsInput) {
          cart {
            entityId
          }
        }
      }
    }
    ```

    ```json filename="GraphQL variables" showLineNumbers copy 
    {
      "addCartLineItemsInput": {

>         "cartEntityId": "XXX",

Reviewer: ```suggestion
        "cartEntityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d",
```

        "data": {
          "lineItems": [
            {
              "quantity": 1,
              "productEntityId": 107
            }
          ]
        }
      }
    }
    ```
  </Tab>
  <Tab>
    ```json filename="Example mutation: Add cart line item" showLineNumbers copy
    {
      "data": {
        "cart": {
          "addCartLineItems": {
            "cart": {
              "entityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d"
            }
          }
        }
      }
    }
    ```
  </Tab>
</Tabs>

### Delete cart

<Tabs items={['Request', 'Response']}>
  <Tab>
    ```graphql filename="Example mutation: Delete cart" showLineNumbers copy
    # Delete your cart in the GraphQL Playground, which will enable you to create a new cart.

>     # Update the GQL variables with details from your cart first, or it will fail.

Reviewer: ```suggestion
    # Add your cart's ID to the GraphQL variables, or the mutation will fail.
```

    
    mutation deleteCart($deleteCartInput: DeleteCartInput!) {
      cart {
        deleteCart(input: $deleteCartInput) {
          deletedCartEntityId
        }
      }
    }
    ```

    ```json filename="GraphQL variables" showLineNumbers copy 
    {
      "deleteCartInput": {

>         "cartEntityId": "XXX"

Reviewer: ```suggestion
        "cartEntityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d"
```

            "createdAt": {
              "utc": "2023-06-01T17:45:23Z"
            },
            "updatedAt": {
              "utc": "2023-06-01T19:07:15Z"
            },
            "customerMessage": "",
            "outstandingBalance": {
              "value": 59.95,
              "currencyCode": "USD"
            },
            "coupons": [],
            "promotions": []
          }
        }
      }
    }
    ```
  </Tab>
</Tabs>

### Add checkout billing address

<Tabs items={['Request', 'Response']}>
  <Tab>
    ```graphql filename="Example mutation: Add checkout billing address" showLineNumbers copy
    mutation addCheckoutBillingAddress($addCheckoutBillingAddressInput: AddCheckoutBillingAddressInput!) {
      checkout {
        addCheckoutBillingAddress(input: $addCheckoutBillingAddressInput) {
          checkout {
            entityId
          }
        }
      }
    }   
    ```

    ```json filename="GraphQL variables" showLineNumbers copy
    {
      "addCheckoutBillingAddressInput": {
        "checkoutEntityId": "812ece1a-da23-46eb-ab6b-c2ee210aae54",
        "data": {
          "address": {
            "firstName": "Joe",
            "lastName": "Blogs",

>             "email": "gqlbeta123@bigcommerce.com",

Reviewer: ```suggestion
            "email": "joe@example.com",
```

use example.com for example URLs

>             "company": "BC",

Reviewer: ```suggestion
            "company": "BlogIndustries",
```

            "address1": "123 Yar st",
            "address2": "",
            "city": "RWC",
            "stateOrProvince": "CA",
            "stateOrProvinceCode": "CA",
            "countryCode": "US",
            "postalCode": "94061",
            "phone": "6501231212",
            "shouldSaveAddress": false
          }
        }
      }
    }
    ```
  </Tab>
  <Tab>
    ```json filename="Example mutation: Add checkout billing address" showLineNumbers copy
    {
      "data": {
        "checkout": {
          "addCheckoutBillingAddress": {
            "checkout": {

>               "entityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d"

Reviewer: ```suggestion
              "entityId": "812ece1a-da23-46eb-ab6b-c2ee210aae54"
```

I believe that the checkout entityId is just the checkout ID. If that's right, this should match the checkout entityId from the request.

            }
          }
        }
      }
    }
    ```
  </Tab>
</Tabs>

### Add a checkout shipping consignment

<Tabs items={['Request', 'Response']}>
  <Tab>
    ```graphql filename="Example mutation: Add a checkout shipping consignment" showLineNumbers copy
    mutation addCheckoutShippingConsignments($addCheckoutShippingConsignmentsInput: AddCheckoutShippingConsignmentsInput!) {
      checkout {
        addCheckoutShippingConsignments(input: $addCheckoutShippingConsignmentsInput) {
          checkout {
            entityId
            shippingConsignments {
              entityId
              availableShippingOptions {
                entityId
              }
              selectedShippingOption {
                entityId
              }
            }
          }
        }
      }
    }
    ```

    ```json filename="GraphQL variables" showLineNumbers copy
    {
      "addCheckoutShippingConsignmentsInput": {
        "checkoutEntityId": "812ece1a-da23-46eb-ab6b-c2ee210aae54",
        "data": {
          "consignments": [
            {
              "address": {
                "firstName": "Joe",
                "lastName": "Blogs",

>                 "email": "gqlbeta123@bigcommerce.com",

Reviewer: ```suggestion
                "email": "joe@example.com",
```

>                 "company": "BC",

Reviewer: ```suggestion
                "company": "BlogIndustries",
```

                "address1": "123 Yar st",
                "city": "RWC",
                "stateOrProvince": "CA",
                "stateOrProvinceCode": "CA",
                "countryCode": "US",
                "postalCode": "94061",
                "phone": "6501231212",
                "shouldSaveAddress": false
              },
              "lineItems": [
                {
                  "lineItemEntityId": "0a281902-d548-4d2e-a930-dc2aae9bfc39",
                  "quantity": 1
                }
              ]
            }
          ]
        }
      }
    }
    ```
  </Tab>
  <Tab>
    ```json filename="Example mutation: Add a checkout shipping consignment" showLineNumbers copy
    {
      "data": {
        "checkout": {
          "addCheckoutShippingConsignments": {
            "checkout": {

>               "entityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d",

Reviewer: ```suggestion
              "entityId": "812ece1a-da23-46eb-ab6b-c2ee210aae54",
```

    {
      "completeCheckoutInput": {
        "checkoutEntityId": "812ece1a-da23-46eb-ab6b-c2ee210aae54"
      }
    }
    ```
  </Tab>
  <Tab>
    ```json filename="Example mutation: Complete checkout" showLineNumbers copy
      {
        "data": {
          "checkout": {
            "completeCheckout": {
              "orderEntityId": 106,
              "paymentAccessToken": "eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE2ODU3MjQwMzQsIm5iZiI6MTY4NTcyMDQzNCwiaXNzIjoicGF5bWVudHMuYmlnY29tbWVyY2UuY29tIiwic3ViIjoidmpid3FiYWJwMSIsImp0aSI6IjAzODU3ODk2LTdkY2YtNDIzNi04OTQ5LWU0MjcyYWU3ZGZjMSIsImlhdCI6MTY4NTcyMDQzNCwiZGF0YSI6eyJzdG9yZV9pZCI6IjEwMDI4ODA3NDYiLCJvcmRlcl9pZCI6IjEwNiIsImFtb3VudCI6MjUwMCwiY3VycmVuY3kiOiJVU0QifX0.iiJ96cYKtk2-oLRXvZHs1lWUl9v8JkEkCdHShbyfEK4"
            }
          }
        }
      }
    ```
  </Tab>
</Tabs>

## Giving feedback

We invite you to give feedback on the GraphQL Storefront schema and the capabilities of the API to let us know if it meets your needs. In particular, we want to know about anything that seems confusing, or anything that you think is missing. We want to ensure that this API meets your needs, and we're open to adding features.

> Join our [Developer Community](/community) to share your feedback with us in the BigCommerceDevs Slack or our Discord server.

Reviewer: ```suggestion
Join our [Developer Community](/community) to share your feedback with us in the BigCommerceDevs Slack or on our Discord server.
```

# GraphQL Storefront API: Carts and Checkout

<Callout type="info">
  While the GraphQL Storefront API is feature-complete, the carts and checkouts schemas are in **beta**. We may introduce changes without notice.
</Callout>

BigCommerce's GraphQL Storefront API provides the same access to cart and checkout objects as the REST Storefront API. Having access to cart and checkout data in the graph alongside related data makes it easier to build headless storefront applications. 

Developers can use frontend GraphQL to create an end-to-end shopper experience. Cart and checkout-related mutations and queries can do the following:
* Add products to a cart
* View or manage a cart
* Fill out checkout details
* Determine the cost of shipping and select a shipping method

> * Complete checkout by converting a cart into an order

Reviewer: ```suggestion
* Complete checkout by converting a cart into an order
* Generate a payment access token
```

    {
      "data": {
        "site": {
          "cart": {
            "entityId": "bb916deb-ddd6-4586-b65b-b8385e0e7a9d",
            "currencyCode": "USD",
            "isTaxIncluded": false,
            "baseAmount": {
              "currencyCode": "USD",
              "value": 25
            },
            "discountedAmount": {
              "currencyCode": "USD",
              "value": 0
            },
            "amount": {
              "currencyCode": "USD",
              "value": 25
            },
            "discounts": [
              {
                "entityId": "50892aad-8e57-4a86-9668-c1f1ee493cff",
                "discountedAmount": {
                  "currencyCode": "USD",
                  "value": 0
                }
              }
            ],
            "lineItems": {
              "physicalItems": [
                {
                  "entityId": "50892aad-8e57-4a86-9668-c1f1ee493cff",
                  "parentEntityId": null,
                  "variantEntityId": 74,
                  "productEntityId": 111,
                  "sku": "SM13",
                  "name": "[Sample] Smith Journal 13",

>                   "url": "https://cart-graphql-playground.mybigcommerce.com/url/",

Reviewer: ```suggestion
                  "url": "https://{store_hash}.mybigcommerce.com/{product-slug}/",
```

>                   "imageUrl": "https://cdn11.bigcommerce.com/s-xxxxx/products/111/images/300/journal.jpg",

Reviewer: ```suggestion
                  "imageUrl": "https://cdn11.bigcommerce.com/s-{store_hash}/products/111/images/300/journal.jpg",
```
